src/uct/ib/ud/base/ud_iface.h

```diff
@@ -131,6 +135,48 @@ typedef struct uct_ud_iface_ops {
                                            struct ibv_qp **qp_p);
 } uct_ud_iface_ops_t;
 
+/* device GIDs set */
+KHASH_TYPE(uct_ud_iface_gid, union ibv_gid, char);
+
+static UCS_F_ALWAYS_INLINE
+khint32_t uct_ud_iface_kh_gid_hash_func(union ibv_gid gid)
+{
+    long len = uct_ib_gid_calc_len(&gid);
+    uint32_t *gid32_p;
+
+    if (len == UCS_IPV4_ADDR_LEN) {
+        UCS_STATIC_ASSERT(UCS_IPV4_ADDR_LEN == sizeof(uint32_t));
+        gid32_p = (uint32_t*)UCS_PTR_BYTE_OFFSET(&gid + 1, -len);
+        return kh_int_hash_func(*gid32_p);
+    } else {
+        ucs_assert(len == UCS_IPV6_ADDR_LEN);
+        UCS_STATIC_ASSERT(UCS_IPV6_ADDR_LEN == sizeof(gid.global));
+        return kh_int64_hash_func(gid.global.subnet_prefix ^
+                                  gid.global.interface_id);
+    }
```

Comment:
i think it would be simpler and faster to keep gids with constant size. if we have short ipv4 gid we can pad it with 0-s

Reply:
unhandled

Reply:
handled in this commit https://github.com/openucx/ucx/pull/5237/commits/c1f1c6f99d4793a9fa77a838e3f3ca53e19b09f2